src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -2357,3 +2365,1577 @@ IsForeignTable(Oid relationId)
+			{
+				counts.cancelled++;
+				break;
+			}
+
+			default:
+			{
+				elog(ERROR, "unknown state in pg_dist_background_tasks");
+			}
+		}
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgDistBackgroundTasks, NoLock);
+
+	return counts;
+}
+
+
+/*
+ * UpdateBackgroundJob updates the job's metadata based on the most recent status of all
+ * its associated tasks.
+ *
+ * Since the state of a job is a function of the state of all associated tasks this
+ * function projects the tasks states into the job's state.
+ *
+ * When Citus makes a change to any of the tasks associated with the job it should call
+ * this function to correctly project the task updates onto the jobs metadata.
+ */
+void
+UpdateBackgroundJob(int64 jobId)
+{
+	JobTaskStatusCounts counts = JobTasksStatusCount(jobId);
+	BackgroundJobStatus status = BACKGROUND_JOB_STATUS_RUNNING;
+
+	if (counts.cancelled > 0)
+	{
+		status = BACKGROUND_JOB_STATUS_CANCELLED;
+	}
+	else if (counts.blocked + counts.runnable + counts.running + counts.error +
+			 counts.unscheduled == 0)
+	{
+		/* all tasks are done, job is finished */
+		status = BACKGROUND_JOB_STATUS_FINISHED;
+	}
+	else if (counts.error + counts.unscheduled > 0)
+	{
+		/* we are either failing, or failed */
+		if (counts.blocked + counts.runnable + counts.running > 0)
+		{
+			/* failing, as there are still tasks to be run */
+			status = BACKGROUND_JOB_STATUS_FAILING;
+		}
+		else
+		{
+			status = BACKGROUND_JOB_STATUS_FAILED;
+		}
+	}
+	else if (counts.blocked + counts.runnable + counts.running > 0)
+	{
+		status = BACKGROUND_JOB_STATUS_RUNNING;
+	}
+	else
+	{
+		return;
+	}
+
+	Relation pgDistBackgroundJobs =
+		table_open(DistBackgroundJobsRelationId(), RowExclusiveLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistBackgroundJobs);
+
+	ScanKeyData scanKey[1] = { 0 };
+	const bool indexOK = true;
+
+	/* WHERE job_id = $task->jobId */
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_background_jobs_job_id,
+				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(jobId));
+
+	SysScanDesc scanDescriptor =
+		systable_beginscan(pgDistBackgroundJobs,
+						   DistBackgroundJobsPKeyIndexId(),
+						   indexOK, NULL, lengthof(scanKey), scanKey);
+
+	HeapTuple heapTuple = systable_getnext(scanDescriptor);
+	if (!HeapTupleIsValid(heapTuple))
+	{
+		ereport(ERROR, (errmsg("could not find background jobs entry for job_id: "
+							   UINT64_FORMAT, jobId)));
+	}
+
+	Datum values[Natts_pg_dist_background_tasks] = { 0 };
+	bool isnull[Natts_pg_dist_background_tasks] = { 0 };
+	bool replace[Natts_pg_dist_background_tasks] = { 0 };
+
+	heap_deform_tuple(heapTuple, tupleDescriptor, values, isnull);
+
+	bool updated = false;
+
+#define UPDATE_FIELD(field, newNull, newValue) \
+	{ \
+		replace[((field) - 1)] = (((newNull) != isnull[((field) - 1)]) \
+								  || (values[((field) - 1)] != (newValue))); \
+		isnull[((field) - 1)] = (newNull); \
+		values[((field) - 1)] = (newValue); \
+		updated |= replace[((field) - 1)]; \
+	}
```

Comment:
I think this would be almost as nice to use if it was a function. And that would allow us to actually reuse it for other things. Something like:
```
updated |= SetFieldValue(Anum_pg_dist_background_jobs_started_at, nulls, replace, values, newValue)
updated |= SetFieldNull(Anum_pg_dist_background_jobs_started_at, nulls, replace)
```